File: 2021/Rust/src/day04.rs
use std::time::Instant;

use eyre::Result;
use ndarray::Array2;

#[derive(Debug, Clone)]
struct Board(Array2<Entry>);

impl Board {
    fn is_finished(&self) -> bool {
        for row in self.0.rows() {
            if row.iter().all(|Entry { enabled, number: _ }| *enabled) {
                return true;
            }
        }
        for col in self.0.columns() {
            if col.iter().all(|Entry { enabled, number: _ }| *enabled) {
                return true;
            }
        }
        false
    }
}

#[derive(Debug, Clone)]
struct Entry {
    enabled: bool,
    number: u64,
}

type Parsed = (Vec<u64>, Vec<Board>);

fn parse(input: &str) -> Result<Parsed> {
    let numbers_line = input.lines().next().unwrap();
    let numbers = numbers_line
        .split(',')
        .map(|x| x.parse())
        .collect::<Result<Vec<u64>, _>>()?;

    let boards = input
        .lines()
        .skip(1)
        .collect::<Vec<_>>()
        .split(|x| x.is_empty())
        .filter(|x| !x.is_empty())
        .map(|x| x.to_vec())
        .map(|board| {
            board
                .into_iter()
                .flat_map(|x| x.split_whitespace())
                .map(|x| {
                    x.parse().map(|x| Entry {
                        enabled: false,
                        number: x,
                    })
                })
                .collect::<Result<Vec<Entry>, _>>()
                .map(|x| Board(Array2::from_shape_vec((5, 5), x).unwrap()))
        })
        .collect::<Result<Vec<Board>, _>>()?;
    Ok((numbers, boards))
}

fn apply_number(number: u64, board: &mut Board) {
    for Entry { enabled, number: i } in &mut board.0 {
        if *i == number {
            *enabled = true;
        }
    }
}

fn part1((numbers, boards): &Parsed) -> Result<u64> {
    let mut boards = boards.clone();

    for &number in numbers {
        for board in &mut boards {
            apply_number(number, board);
            if board.is_finished() {
                return Ok(board
                    .0
                    .iter()
                    .filter_map(
                        |&Entry { enabled, number }| if enabled { None } else { Some(number) },
                    )
                    .sum::<u64>()
                    * number);
            }
        }
    }
    unreachable!("At least one board should win")
}

fn part2((numbers, boards): &Parsed) -> u64 {
    let mut boards = boards.clone();
    for &number in numbers {
        for b in &mut boards {
            apply_number(number, b);
        }
        if boards.len() == 1 && boards[0].is_finished() {
            return boards[0]
                .0
                .iter()
                .filter_map(|&Entry { enabled, number }| if enabled { None } else { Some(number) })
                .sum::<u64>()
                * number;
        }
        boards.retain(|x| !x.is_finished());
    }
    unreachable!("All boards should win")
}

pub fn run(input: String) -> Result<()> {
    let parsed = parse(input.as_str())?;
    let start = Instant::now();
    let part1 = part1(&parsed)?;
    let p1_time = Instant::now() - start;
    println!(
        "The solution to 2021 day 03 (part 1) is {}. Took {:?}",
        part1, p1_time
    );
    let start = Instant::now();
    let part2 = part2(&parsed);
    let p2_time = Instant::now() - start;
    println!(
        "The solution to 2021 day 03 (part 2) is {}. Took {:?}",
        part2, p2_time
    );

    Ok(())
}

File: 2021/Haskell/src/Day04.hs
{-# LANGUAGE OverloadedStrings #-}

module Day04
  ( day04,
  )
where

import AoC
import Common
import Control.Monad (join)
import Data.IntSet (IntSet)
import qualified Data.IntSet as IS
import Data.List (find, transpose, (\\))
import Data.Text (Text)
import qualified Data.Text as Text
import Data.Vector (Vector)
import qualified Data.Vector as Vector
import Text.Megaparsec
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L

type Parsed = ([Int], [Board])

data Board = Board [[Int]] IntSet
  deriving (Show)

pInput :: Parser Parsed
pInput = do
  numbers <- pNumbersLine
  _ <- newline >> newline
  boards <- some do
    b <- pBoard
    _ <- many newline
    return b
  return (numbers, boards)

pNumbersLine :: Parser [Int]
pNumbersLine = sepBy L.decimal (char ',')

pRow :: Parser [Int]
pRow = count 5 do
  _ <- many $ char ' '
  num <- L.decimal
  _ <- many $ char ' '
  return num

pBoard :: Parser Board
pBoard = do
  ints <-
    count 5 do
      x <- pRow
      _ <- newline
      return x
  return $ Board ints mempty

hasWon :: Board -> Bool
hasWon (Board board seen) =
  let rows = board
      cols = transpose board
      r = any (null . (\\ IS.toList seen)) rows
      c = any (null . (\\ IS.toList seen)) cols
   in r || c

applyNumber :: Int -> Board -> Board
applyNumber number (Board board seen) = Board board (IS.insert number seen)

getUnseen :: Board -> [Int]
getUnseen (Board ints seen) = filter (\x -> not $ IS.member x seen) (join ints)

getFirstWinner :: [Int] -> [Board] -> Int
getFirstWinner (number : numbers) boards =
  let boards' = applyNumber number <$> boards
      winner = find hasWon boards'
   in case winner of
        Nothing -> getFirstWinner numbers boards'
        Just winner -> number * sum (getUnseen winner)
getFirstWinner [] _ = unreachable "At least one board should win"

getLastWinner :: [Int] -> [Board] -> Int
getLastWinner (number : numbers) boards =
  let boards' = applyNumber number <$> boards
      yetToWin = filter (not . hasWon) boards'
   in case boards' of
        [winner] | hasWon winner -> number * sum (getUnseen winner)
        _ -> getLastWinner numbers yetToWin
getLastWinner [] _ = unreachable "All boards should win"

day04 :: Runner Parsed Int
day04 =
  let year = 2021
      day = 4
      parser :: Text -> Maybe Parsed
      parser = unwrapParser . parse pInput "input"
      part1 (numbers, boards) = return $ getFirstWinner numbers boards
      part2 (numbers, boards) = return $ getLastWinner numbers boards
   in Runner {..}
